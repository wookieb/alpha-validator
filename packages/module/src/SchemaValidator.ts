import {SchemaValidation} from "alpha-validator";

export interface SchemaValidator<T = any> {
    validate: SchemaValidation<T>
}