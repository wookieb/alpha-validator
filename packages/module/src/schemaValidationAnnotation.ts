import * as is from 'predicates';

/**
 * Annotation that marks given service as validation function for given schema names
 */
export function schemaValidationAnnotation(...schemaNames: string[]): Annotation {
    is.assert(is.notEmptyArray, 'You need to provide at least one schema name')(schemaNames);
    is.assert(is.arrayOf(is.notBlank), 'Every schema name must be a non-blank string')(schemaNames);
    return {
        name: NAME,
        schemaNames
    };
}

const NAME = 'alpha-validator/schema-validation';

export interface Annotation {
    name: string;
    schemaNames: string[];
}

export const PREDICATE = is.prop('name', is.eq(NAME));