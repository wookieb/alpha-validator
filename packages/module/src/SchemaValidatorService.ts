import {Annotation, onActivation, Service, ServiceName} from "alpha-dic";
import {schemaValidationAnnotation} from "./schemaValidationAnnotation";
import {SchemaValidator} from "./SchemaValidator";
import * as is from 'predicates';

const assertIsValidator = is.assert(
    is.prop('validate', is.func),
    'Schema validation service needs to be an object with `validate` function. Make sure that SchemaValidator interface is implemented.'
);

/**
 * Annotation that marks service as validation for given schema names.
 * Object has to implement SchemaValidator interface.
 */
export function SchemaValidatorService(schemaNames: string[], serviceName?: ServiceName) {
    return function (clazz: { new(...args: any[]): any }) {
        Service(serviceName)(clazz);
        Annotation(onActivation((validator: SchemaValidator) => {
            assertIsValidator(validator);
            return validator.validate.bind(validator);
        }))(clazz);
        Annotation(schemaValidationAnnotation(...schemaNames))(clazz);
    }
}