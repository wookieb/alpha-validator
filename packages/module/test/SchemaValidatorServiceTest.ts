import {SchemaValidator} from "@src/SchemaValidator";
import {SchemaValidatorService} from "@src/SchemaValidatorService";
import {Container, createStandard, Service} from "alpha-dic";
import {PREDICATE} from "@src/schemaValidationAnnotation";
import * as sinon from 'sinon';
import {Validation} from "alpha-validator";

describe('SchemaValidatorService', () => {

    let container: Container;

    beforeEach(() => {
        container = createStandard();
    });

    it('registers class', async () => {
        const stub = sinon.stub();
        const RESULT = Validation.Success({});

        @SchemaValidatorService('name')
        class Foo implements SchemaValidator {
            validate(data: any, schemaName: string): Validation<any, any> {
                stub.call(this, data, schemaName);
                return RESULT;
            }
        }

        const validations = await container.getByAnnotation(PREDICATE);

        expect(validations)
            .toHaveLength(1);

        const validation = validations[0];
        expect(validation)
            .toBeInstanceOf(Function);

        expect(validation())
            .toStrictEqual(RESULT);

        sinon.assert.calledOn(stub, sinon.match.instanceOf(Foo));
    });

    it('fails to create a service on an object that does not implement SchemaValidator class', () => {
        @SchemaValidatorService('name')
        class Foo {
        }

        return expect(container.getByAnnotation(PREDICATE))
            .rejects
            .toThrowErrorMatchingSnapshot();
    });
});