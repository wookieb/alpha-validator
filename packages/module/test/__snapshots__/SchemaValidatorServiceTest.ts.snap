// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaValidatorService fails to create a service on an object that does not implement SchemaValidator class 1`] = `"Schema validation service needs to be an object with \`validate\` function. Make sure that SchemaValidator interface is implemented."`;
